File: JS/LeetCode热题HOT100/55_跳跃游戏.js
// var canJump = function(nums) {
//     function isFinish(index, maxStep){
//         if(index >= nums.length-1 || nums[index]+index >= nums.length-1) return true
//         maxStep = 0
//         let newIndex = index
//         for(let j=1; j<=nums[index]; j++){
//             if(nums[index+j]+j >= maxStep) {
//                 newIndex = index+j
//                 maxStep = nums[index+j] + j
//             }
//         }
//         maxStep = maxStep - (newIndex - index)
//         if(maxStep === 0 && index < nums.length-1) return false
//         return isFinish(newIndex, maxStep)
//     }
//     return isFinish(0, nums[0])
// };
var canJump = function(nums) {
    if(nums.length === 1) return true
    let cover = 0
    for(let i = 0; i <= cover; i++) {
        cover = Math.max(cover, i + nums[i])
        if(cover >= nums.length - 1) {
            return true
        }
    }
    return false
};
let nums = [1,1,2,2,0,1,1]
console.log(canJump(nums))

File: Python/面试经典150题/数组、目录/55_跳跃游戏.py
from typing import List
# class Solution:
#     def canJump(self, nums: List[int]) -> bool:
#         index = 0
#         while index < len(nums)-1:
#             if nums[index] == 0:
#                 return False
#             far = 0
#             for i in range(nums[index]):
#                 if index+i+1 > len(nums)-1:
#                     break
#                 step = nums[index+i+1] + i
#                 if nums[index+i+1] + i + 1 > far:
#                     far =  step
#                     next_index = index+i+1
#             index = next_index
#         return True

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n, rightmost = len(nums), 0
        for i in range(n):
            if i <= rightmost:
                rightmost = max(rightmost, i + nums[i])
                if rightmost >= n - 1:
                    return True
        return False


nums = [2,0]
print(Solution().canJump(nums))



